Step3 추가사항

beemiel opened the pull request:
# step3

- 클래스 구조

![클래스 구조1](https://i.imgur.com/wHEwWLg.png)

![클래스 구조2](https://i.imgur.com/nuYniE0.png)

- 요구사항에 따른 추가 사항 및 변경 사항

1. 비행,달리기,에스퍼 세 타입의 클래스를 추가하고 기존의 Moster클래스를 추상클래스로 만들어서 상속받음. 
2. enum을 사용하여 입력받은 타입을 검사하고 타입에 맞는 객체를 생성해줌.
3. 모든 라운드가 종료된 후 최종 우승자를 계산한 후 출력



- 클래스 상세
  - Monster Class
    - static final int RUN_NUMBER_OF_RUNNING - 달리기 타입 몬스터의 움직임을 결정하는 기준점
    - static final int RUN_NUMBER_OF_FLYING - 비행타입 몬스터의 움직임을 결정하는 기준점
    - static final int RUN_NUMBER_OF_ESPER - 에스퍼타입 몬스터의 움직임을 결정하는 기준점
    - String name - 몬스터의 이름
    - String type - 몬스터의 타입
    - int isRunNumber - 난수가 true인 조건
    - String totalMove - 총 걸음 수
    - Random random - random 재활용
    - StringBuffer stringBuffer - 문자열 추가
    - getName() - 몬스터의 이름 반환
    - setName(String name) - 몬스터의 이름 저장
    - isMove() -  isRunNumber 보다 크거나 같을 경우 true 반환
    - moveOrStop() - 추상메소드
    - run(int count) - 입력받은 카운트 수만큼 반복
    - toString() - 몬스터의 이름,타입,총 걸음 수 출력
  - Running Class
    - moveOrStop() - isMove()의 값이 true일 경우 1 추가
  - Flying Class
    - moveOrStop() - isMove()의 값이 true일 경우 3 추가
  - Psychic Class
    - moveByRandom() - 움직임 수를 난수로 생성 후 반환
    - moveOrStop() - isMove()의 값이 true일 경우 moveByRandom()만큼 추가
  - Message Class
    - 사용자에게 출력할 상수 문자열들의 집합
  - Race Class
    - ArrayList<Monster> monsters - 입력받은 몬스터 리스트
    - racing() - 레이스 정보를 입력받고 순차적으로 경기를 실행한 후 결과를 출력
    - startRacing(int count) - 카운트만큼 경기를 실행
  - Input Class 
    - Scanner scanner - 사용자에게 입력을 받기 위한 스캐너
    - inputMonster() - 몬스터 수 입력
    - inputMonsterInfo() - 몬스터 정보 입력, 그 정보로 생성된 객체 반환
    - inputCount() - 라운드 수 입력
    - close() - scanner 닫음 
  - MonsterType enum
    - 세 가지 타입 정의
    - findTypeByInput() - 입력한 타입이 존재하는지 확인
    - makeMonsterByType() - 각 타입마다 다른 객체를 생성해서 반환
  - Result Class
    - contenOfGame() - 콘솔에 경기 내용 출력
    - resultOfGame() - 콘솔에 경기 결과 출력 
    - appendString(ArrayList<Monster> winners) - 우승자 리스트에 있는 우승자들의 이름을 하나의 문자열로 반환
  - Calculator Class
    - ArrayList<Monster> winnerMonsters - 우승자 리스트
    - calculateResults() - calculateHigjhRecord()와 addWinnerMonster()를 실행한 후 우승자 리스트 반환
    - calculateHigjhRecord() - 최고점수 계산
    -  addWinnerMonster() - 최고점수와 같으면 우승자 리스트에 추가
  - Setting Class
    - int monster - 몬스터의 수
    - int count - 시도 횟수
    - Input input - 입력받는 객체
    - ArrayList<Monster> monsters - 입력받은 몬스터 리스트
    - getCount() - 시도 횟수 반환
    - getMonsters() - 몬스터 리스트 반환
    - setRace() - 레이스를 위한 사전 정보 입력 메서드들을 실행
    - setMonsterNumber() - 몬스터 객체의 수를 받음
    - setCOunt() - 시도 횟수를 받음
    - StringToInt() - 문자열을 int형으로 변환
    - setMonster() - 몬스터 변수의 수만큼 몬스터를 만듦
    - makeMonster() - 몬스터 정보를 입력받아서 몬스터 객체를 생성한 후 몬스터 리스트에 추가
    - checkInputFormat() - 입력값의 포맷을 확인
    - addMonster() - 몬스터 리스트에 몬스터 객체 추가
    - checkType() - 입력값의 타입을 체크
    - gameClose() - 스캐너 닫음

beemiel committed:
다형성을 위해 Monster클래스를 추상클래스로 변경

totalMove - 몬스터의 총 걸음수
random - random 객체를 재사용하기 위한 객체
stringBuffer - StringBuffer객체를 재사용하기 위한 객체

isMove() - abstract 랜덤값을 조건에 따라 움직임을 boolean으로 반환
moveOrStop() - abstract isMove값에 따라 totalMove에 문자열을 추가
toString() - Object 클래스의 toString()을 override해서 몬스터의 이름,
타입, 총 움직임 수를 문자열로 반환

beemiel committed:
비행 타입 몬스터 클래스 Flying 추가

isMove() - 랜덤값이 6이상이면 true 반환
moveOrStop() - isMove의 값이 true이면 totalMove에 문자열 추가

beemiel committed:
달리기 타입 몬스터 Running 클래스 추가

isMove() - 랜덤값이 4이상이면 true반환
moveOrStop() - isMove()의 값이 true이면 totalMove에 문자열 추가

beemiel committed:
에스퍼 타입 몬스터 Psychic 클래스 추가

isMove() - 랜덤값이 9이상이면 true반환
moveOrStop() - isMove()의 값이 true이면 moveByRandom()을 실행해서
totalMove에 담음
moveByRandom() - 1~100까지의 랜덤값을 구해서 그 수만큼 totalMove에
문자열 추가

beemiel committed:
Input class <feat>: 이름과 타입을 입력받는 메소드 추가

inputMonsterInfo() - 몬스터의 이름과 타입을 사용자가 결정할 수 있는
기능이 주가되어서 사용자에게 스캐너로 몬스터의 이름과 타입을 입력받음
message 객체 - output 클래스의 정의를 메세지를 출력하는 것이 아니라
게임의 결과를 출력하는 것으로 변경했음. 그리고 Output에 있던 메소드들을
Message 클래스에 옮겼음. 그래서 기존의 output객체를 message로 변경함

beemiel committed:
Output<feat>: 게임 내용 출력으로 클래스 정의 변경

기존에는 message들을 콘솔에 출력하는 용도로 사용하던 클래스였으나
게임의 내용을 출력하는 클래스로 변경

contentOfGame() - 몬스터들이 움직인 횟수를 콘솔창에 출력해줌
resultOfGame() - 게임이 끝나고 우승자를 콘솔창에 출력해줌

beemiel committed:
Message <feat>: 콘솔창에 메세지들을 출력하도록 추가

기존에 Output class에 있던 기능들을 가져와서 Message class에 추가하였음
모두 메세지를 콘솔창에 출력하는 메소드들로 구성되어 있음

beemiel committed:
Monster<feat>: 몬스터의 이름을 얻는 get 추가

추상메소드에 주석을 달아서 어떤 기능인지 설명한 내용 추가

getName() - 게임이 끝난 후 최종 우승자를 선별할 때 우승한 몬스터의이름을
출력하기 위해서 몬스터의 이름을 반환해주는 get메소드를 추가함

beemiel committed:
Running<modified>: StringBuffer를 사용하는 부분 수정

stringBuffer를 부모클래스인 Monster의 멤버 변수에 추가함.
그래서 생성해줌

beemiel committed:
Flying<modified>: StringBuffer 수정

StringBuffer를 생성해주는 부분 추가

beemiel committed:
Psychic <modified>: StringBuffer 객체 생성

문자열을 붙이기 위해서 StringBuffer 객체를 생성함

beemiel committed:
Race <feat>: 최종 우승자를 가려내는 로직 추가

winnerMonsters - 우승자를 담을 ArrayList
message - 콘솔에 message를 출력
input - 콘솔에서 사용자에게 입력을 받아 처리
output - 콘솔에 게임의 결과를 출력

racing() - 기존에 output을 message객체로 변경함
우승자를 계산하는 메소드 추가
makeMonster() - 입력받은 count만큼 반복문을 이용하여
사용자로부터몬스터의
이름과 타입을 입력 받는다. 이 정보들을 통해 몬스터 객체를 생성하여
몬스터를 저장하는 ArrayList에 저장
returnMonsterType() - 사용자가 입력한 타입에 따라서 다른 클래스의 객체를
반환
calculateResults() - 최고 기록을 가져온 다음 몬스터 리스트에 있는
몬스터들의 기록과 대조해서 최고 기록과 같으면 우승자를 담는 winner
리스트에 담음
addWinnerMonster(Monster monster, int highReord) - 매개변수로 받은 최고
기록과
몬스터의기록을 비교하는 로직을 조건문으로 작성하여 참일 경우에는우승자
리스트에
해당 몬스터를 추가함
calculateHighRecord() - 우승자를 선정하기 위해서 최고 기록을 찾는 로직
구현
모든 몬스터 리스트를 반복문으로 돌리면서 기존에 저장된 기록과 비교할
몬스터의
기록 중 큰수를 record에 저장함
이를 모든 리스트와 모두 대조함. 그 후 최고 기록을 반환해줌.

beemiel committed:
<new>: 몬스터의 타입을 구분하기 위해  enum 추가

각 타입에 맞는 객체를 생성하기 위해서 enum을 추가하여 사용함
Generator() - 각 타입별 객체를 생성
getMonster() - 갖고 있는 monster객체를 반환:

beemiel committed:
Running <modified>: 생성자 수정, isMove() 삭제

Running(String type) - enum을 사용하면서 기존의 생성자를 수정함
isMove() - 추상클래스인 Monster에 isRunNumber라는 멤버변수를 추가해서
isMove()를 각 클래스마다 구현하지 않고 Monster클래스에서 구현해서
상속을 받도록 수정함

beemiel committed:
Flying <modified>: 생성자 수정, isMove()삭제

Flying(String type) - enum에서 매개변수를 type만으로 객체를 생성하기
때문에 생성자 수정
isMove() - 부모 클래스인 Monster 클래스에서 구현하므로 삭제함

beemiel committed:
Psychic <modified>: 생성자 수정, isMove()삭제

Psychic(String type) - enum사용으로 인해 생성자 수정
isMove() - 부모클래스인 Monster클래스에서 구현하므로 삭제

beemiel committed:
Message <modified>: MONSTER_INFO 문자열 수정

기존의 문자열에서 예시를 예> 크롱,FLY 로 변경

beemiel committed:
MonsterType <modified>: 변수명 수정

첫글자만 대문자인 것을 모두 대문자로 수정

beemiel committed:
Monster <modified>: 멤버 변수 추가, isMove(),setName()구현

isRunNumber - 각 타입마다 isMove()가 true가 될 확률이 다르기 때문에
isMove()를 Monster 클래스에서 구현하기 위해서 멤버 변수로 빼서
재사용하게 함
setName() - 몬스터 객체 생성시 이름을 지정해주기 위한 set메소드
isMove() - 기존에는 추상메소드였지만 isRunNumber를 필드로 빼서
Monster클래스에서
바로 구현하도록 수정함

beemiel committed:
Race <modified>: 예외처리 추가, makeMonster() 수정

예외처리 - 사용자의 입력을 받는 부분에서 발행하는 예외를 처리하여
제대로 입력될 때까지 반복하여 입력을 받는다.
makeMonster() - input.inpputMonsterInfo()에서문자열 배열을 반환받지 않고
입력한 타입에 따라 다른 monster객체를 반환받아서 리스트에 넣도록 수정함

beemiel committed:
Output <modified>: 출력 메세지 추가, StringJoiner 사용

외부에 있던 출력문을 Output클래스 안으로 가져옴
resultOfGame() - 기존에 StringBuffer를 사용하던 부분을 StringJoiner로
변경
앞뒤와 중간에 알아서 문자를 붙여줘서 코드가 더 간결하고 편리해짐

beemiel committed:
Input <modified>: 예외 던지기 추가, inputMonsterInfo() 수정

inputMonster() - NumberFormatException 을 던져주도록 추가
inputCount() - InputMismatchException을 던져주도록 추가
inputMonsterInfo() - 입력받은 값을 통으로 enum.valueOf()를 사용하여
enum에 선언된 타입일 경우, 해당 타입의 enum을 반환한 후 name을
저장해준 다음
해당 객체를 getMonster()로 반환받아서 다시 반환해줌.
입력한 타입이 존재하지 않을 경우에는 예외를 반환하여 알림 메세지를
출력해주고
다시 입력하도록 반복문을 돔

beemiel committed:
README <add>: 스텝3 내용 추가

스텝3에 관한 내용 추가

beemiel committed:
README <modified>: 이미지 수정

이미지 수정

honux77 commented on src/main/java/com/beemiel/monsterrace/Flying.java:
```diff
@@ -0,0 +1,18 @@
+public class Flying extends Monster {
+
+    public Flying(String type){
+        super(type,6);
```
6같은 매직 넘버는 코드 가독성을 떨어뜨립니다. 상수를 하나 만들어서 이용해 주세요.

honux77 commented on src/main/java/com/beemiel/monsterrace/Flying.java:
```diff
@@ -0,0 +1,18 @@
+    public void moveOrStop() {
+        if(isMove()){
+            stringBuffer = new StringBuffer(totalMove);
+            stringBuffer.append("▷▷▷");
```
ㅋㅋㅋㅋㅋㅋ 귀엽네요.

honux77 commented on src/main/java/com/beemiel/monsterrace/Input.java:
```diff
@@ -1,19 +1,38 @@
+                MonsterType monsterType = MonsterType.valueOf(type);
+                monsterType.getMonster().setName(name);
+                return monsterType.getMonster();
+            } catch (Exception e){
```
Exception 말고 구체적인 exception을 처리해 주는게 좋아요.

honux77 commented on src/main/java/com/beemiel/monsterrace/Input.java:
```diff
@@ -1,19 +1,38 @@
-        output.gameStart();
-        output.getMonster();
-        return scanner.nextInt();
+    public int inputMonster() throws NumberFormatException{
```
이 메소드가 입력과 출력을 담당하고 있어서, 이름보다 많은 일을 하는 느낌이에요. 클래스도 마찬가지고
message가 있긴 하지만 InputView 의 역할을 하고 있네요.
메소드 안에서 message읨 메소드를 출력하는 것보다 상위 객체가 하는게 더 좋지 않을까 생각이 듭니다.

객체의 메소드는 한가지 일을 하도록 하는게 좋아요.

honux77 commented on src/main/java/com/beemiel/monsterrace/Input.java:
```diff
@@ -1,19 +1,38 @@
+                String type = info[1];
+                String name = info[0];
+                MonsterType monsterType = MonsterType.valueOf(type);
+                monsterType.getMonster().setName(name);
```
아직 monsterType을 보지 않았는데 이 부분도 해당 코드만으로는 조금 동작의 유추가 어렵네요. 가능하면 현재 코드만으로 이해가 되도록 짜면 좋을 것 같아요.

honux77 commented on src/main/java/com/beemiel/monsterrace/Message.java:
```diff
@@ -1,10 +1,41 @@
-    final static String RESULT = "<결  과>";
-    final static String CLOSE = "<경기 종료>";
-    final static String RUN = "▷";
+
```
static 클래스에 static 메소드를 사용해도 좋을 것 같습니다. 상태가 없으니까요.

honux77 commented on README.md:
```diff
@@ -1,3 +1,71 @@
+    - run(int count) - 입력받은 카운트 수만큼 반복
+    - toString() - 몬스터의 이름,타입,총 걸음 수 출력
+  - Running Class
+    - moveOrStop() - isMove()의 값이 true일 경우 1 추가
```
1과 3은 뭘까요?

honux77 commented on src/main/java/com/beemiel/monsterrace/MonsterType.java:
```diff
@@ -0,0 +1,16 @@
+package com.beemiel.monsterrace;
+
+public enum MonsterType {
```
enum의 선언과 사용이 조금 이상한 듯해요. 다시 공부해 보고 수정하거나, 아예 enum을 빼 보세요.

honux77 commented on src/main/java/com/beemiel/monsterrace/Message.java:
```diff
@@ -1,10 +1,41 @@
+    final String TITLE = "<MONSTER RACE>";
+    final String START = "\n레이스 시작";
+    final String MONSTER = "몬스터는 모두 몇 마리인가요?";
+    final String MONSTER_INFO = "\n경주할 몬스터 이름과 종류를 입력하세요 (예> 크롱,FLY)";
```
사용자에게 FLY라고 입력을 시키는 건 너무 가혹한데요. 원래의 요구사항대로 비행으로 입력하도록 고치면 좋겠네요.

honux77 replied:
그리고 현재 프로그램이  정상적으로 돌아가지 않아요 ㅜㅜ.

honux77 reviewed (changes requested):
정상적으로 돌아가도록 고쳐 주세요. PR을 다시 보낼 필요는 없고, 이 브랜치에서 작업하고 push만 하면 됩니다.

beemiel committed:
MonsterType<수정>: 상수 재정의, 값 찾는 메서드,몬스터 만드는 메서드 추가

enum을 잘못 사용하여 수정
기존의 상수들이 가지고 있던 값이 객체였어서 적절한 값으로 변경
findTypeByInput(String input) - 상수 중 매개변수와 동일한 값을 갖고 있는
상수를
반환해준다. 만약 없다면 예외를 반환해줌
makeMonsterByType(String name) - enum객체가 갖고 있는 값에 따라서 다른
타입의
객체를 반환한다.

beemiel committed:
Setting<추가>: 레이스에 필요한 정보 세팅하는 클래스 추가

input으로부터 레이스에 필요한 정보들을 입력받는다
입력 받은 정보들이 적절한 형식과 값을 갖고 있는지 확인하고
필요한 경우 문자열을 int형으로 변환한다.

int monster - 몬스터의 수
int count - 시도 횟수
Input input - 입력값을 받는 객체
ArrayList<Monster> monsters - 몬스터 객체들을 담는 리스트

getCount() - count 반환한다
getMonsters() - 몬스터 리스트 반환한다
setRace() - 정보를 입력받는 메서드를 모두 실행한다
setMonsterNumber() - monster변수의 값을 입력받는다
setCount() - count변수의 값을 입력받는다
StringToInrt() - 입력받은 문자열을 int형으로 변환해서 반환한다
setMonster() - monster변수의 값만큼 몬스터를 만드는 메서드를 실행한다
makeMonster() - 입력받은 문자열을 체크해서 몬스터 객체를 생성한 후
리스트에 추가한다
checkInputFormat() - 입력받은 문자열의 형식을 정규표현식으로 검사한다
addMonster() - 리스트에 몬스터 객체를 추가한다
checkType() - 입력받은 타입이 적절한 타입인지 검사한다
gameClose() - scanner를 닫는다

beemiel committed:
Calculator<추가>: 게임의 결과로 우승자를 계산하는 클래스

winnerMonsters - 우승자 리스트
calculateHighRecord() - 몬스터들의 기록들을 비교해서  최고 기록을
반환한다.
calculateResults() - 최고 기록을 연산하는 메서드와 우승자 리스트에
추가하는
메서드를 실행한 후 우승자 리스트를 반환한다.
addWinnerMonster() - 최고 기록과 몬스터의 기록을 비교해서 최고기록과
같으면
우승자 리스트에 추가한다

beemiel committed:
Input<수정>: 기존의 메서드들을 수정

모든 메서드에서 입력뿐만 아니라 출력까지 수행하는 메서드들에서
출력부분을
제거하고 입력받은 값만 반환하도록 수정

beemiel committed:
Monster<수정>: 상수 추가

몬스터들마다 움직임의 기준이 되는 상수들을 정의

beemiel committed:
Race<수정>: 경기 결과를 계산하는 메서드를  다른 클래스로 분리

기존에 경기 결과를 계산하여 최종 우승자를 찾는 메서드들을 Calculator라는
클래스로 분리함

startRacing() - 실제 게임내용을 연산하는 메서드

beemiel committed:
Result<수정>: Output클래스를 Result 클래스로 변경

contentOfGame() - 게임 결과를 출력한다
appendString() - 우승자들의 이름을 하나의 문자열로 붙여서 반환한다
resultOfGame() - 우승사를 출력한다

beemiel committed:
Message<수정>: 메서드들을 없애고 상수들 집합으로 변경

기존에 있던 출력메서드들을 없애고 문자열 상수들로만 수정

beemiel committed:
Flying<수정>: 매직넘버를 상수로 변경

매직넘버를 사용하던 곳을 상수로 변경

beemiel committed:
Psychic<수정>: 매직넘버를 상수로 변경

매직넘버를 사용하던 곳을 상수로 변경

beemiel committed:
Running<수정>: 매직넘버 상수로 변경

매직넘버를 사용하던 곳을 상수로 변경

beemiel committed:
MonsterType<수정>: 필요없는 출력문 삭제

makeMonsterByType()에 있던 출력문구를 삭제

beemiel committed:
step3 README.md 수정

beemiel commented:
@honux77 수정을 해보았습니다. enum 아직도 아닌가요? 빼는게 나을까요?

honux77 reviewed (approved):
좀 깔끔해졌네요. 타입 찾는 부분을 그렇게까지 해야 하나 싶긴 한데 그래도 괜찮아요 ㅎㅎ
